pkg/credentialprovider/apis/criauth/types.go

```diff
@@ -0,0 +1,159 @@
+	DefaultCacheDuration metav1.Duration
+
+	// extraArgs specifies arguments to be passed to the plugin via environment variables.
+	// +optional
+	ExtraArgs []PluginArg
+}
+
+// PluginArg is used for passing extra arguments to the exec credential plugin binary.
+// Exec plugins expect the arguments to be passed in the format "key=value".
+type PluginArg struct {
+	Name  string
+	Value string
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// CRIAuthPluginRequest includes the image that the kubelet requires authentication for.
+// Kubelet will pass this request object to the plugin via stdin.
+type CRIAuthPluginRequest struct {
+	metav1.TypeMeta
+
+	// image is the container image that is being pulled as part of the
+	// CRI auth plugin request. Plugins may optionally parse the image
+	// to extract any information required to fetch credentials.
+	Image string
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// CRIAuthPluginResponse holds credentials that the kubelet should use for the specified
+// image provided in the original request. Kubelet will read the response from the plugin via stdout.
+type CRIAuthPluginResponse struct {
+	metav1.TypeMeta
+
+	// authEntries is a list of AuthEntry. Each AuthEntry should contain credentials that can be used
+	// for the provided image in CRIAuthPluginRequest. AuthEntry also contains metadata for the kubelet
+	// such as a match key and cache duration.
+	AuthEntries []AuthEntry
+}
+
+// AuthEntry contains credentials that can be used for the provided image in CRIAuthPluinRequest.
+// It also contains useful metadata for the kubelet such a match key and cache duration.
+type AuthEntry struct {
+	// matchKey is the match URL that credentials are valid for. The kubelet will use
+	// this value when looking up images in the auth keyring. This is also the cache key
+	// used by the kubelet.
+	//
+	// matchKey should be a URL string pattern which can optionally contain a port and path.
+	// Globs can be used in the hostname, but not in the port or the path. The kubelet will use
+	// the AuthConfig in this entry for any subsequent images matching against this match key while
+	// this entry is still cached.
+	//
+	// Examples of values:
+	//   - 123456789.dkr.ecr.us-east-1.amazonaws.com
+	//   - *.azurecr.io
+	//   - gcr.io
+	//   - *.*.registry.io
+	//   - registry.io:8080/path
+	MatchKey string
+
+	// cacheDuration indicates the duration the provided credentials should be cached for.
+	// The kubelet will use this field to set the in-memory cache duration for credentials
+	// in the AuthConfig. If nil, the kubelet will use defaultCacheDuration provided in
+	// CRIAuthPluginConfig. If set to 0, the kubelet will not cache the provided AuthConfig.
+	CacheDuration *metav1.Duration
+
+	// authConfig contains authentication information for the container registry.
+	// Only username/password based authentication is supported today, but more authentication
```

Comment:
Will allowing plugins to support additional/alternative authentication methods besides a username/password pair be as simple as adding additional fields to `AuthConfig` and allowing plugins to populate those fields?

Reply:
Yes I think we want to support alternative auth methods in the future, related thread here https://github.com/kubernetes/kubernetes/pull/94196#discussion_r476821486